rfcs/print_test.md

```diff
@@ -0,0 +1,151 @@
+test type").
+
+## Details
+
+This proposal adds a new test type, called `print-reftest`.
+
+This new test type is similar to the existing `reftest` type; it uses
+`<link rel=match>` and `<link rel=mismatch>`, combining in the same
+way as for normal reftests. It also supports `reftest-wait`. Instead
+of comparing screenshots of the viewport, a print-reftest first
+renders the document to a PDF, and then computes the reftest result by
+rasterising the test and ref PDFs at 96 dpi and comparing the results
+in a page-by-page fashion using the same logic as for standard
+reftests. In order to pass an equal comparison test must have all
+pages the same whereas a not-equal comparison test must have at least
+one page different.
+
+This tests are identified by a `print` type-flag in the filename
+(i.e., "foo-print.html").
+
+The paper sizes for the PDF are 5 inches by 3 inches with 0.5 inch
+margins on all sides. Initially this is fixed, but this could later be
+relaxed by adding a new meta property to specify the required
+dimensions.
+
+In some cases it may not be desirable to compare all pages of the
+output. To support this use case a new meta value is used; `<meta
+name="page-ranges">`. The content value of this element follows the
+form:
+```
+content = (RefPath ":")? Range ("," Range)*
+Range = Int | Int "-" Int
+Int = [0-9]+
+RefPath = /* Relative path to a reference of the current test */
+```
+Multiple `<meta name=page-ranges>` elements may be specified in a
+single test, each applying to a particular reference. A range with no
+reference specified applies to the test itself. Ranges specified in
+reference files are an error that are ignored. The `page-ranges`
+specifier for a particular document is interpreted as for the [parse a
+page range](https://www.w3.org/TR/webdriver/#dfn-parse-a-page-range)
+algorithm in the WebDriver spec.
+
+### Implementation in wptrunner
+
+The cross-browser implementation of this feature relies on two things:
+* The [print endpoint in the webdriver
+  standard](https://www.w3.org/TR/webdriver/#print)
+* pdf.js providing a cross-browser mechanism to render a PDF without
+  depending on external tools, many of which are not cross-platform or
+  have inappropriate licensing.
+
+This stack is good enough to prove that the feature can be implemented
+in any browser that has the relevant standards support, and will be
+enough to run the feature in CI and get results on wpt.fyi. Vendors
+may wish to use a custom implementation in their own infrastructure
+for reasons of performance. Indeed if the browser is able to rasterise
+directly to the paginated form that will avoid the need to create an
+intermediate PDF.
+
+## Risks
+
+* Browsers' rendering to PDF is non-deterministic and therefore
+  identical pages won't be guaranteed to yield a PDF that is
+  graphically the same on each load. This can be addressed with
+  appropriate fuzzy specifiers.
+
+* We don't catch non-visual differences in the PDFs, such as
```

Comment:
This risk gets into the dicey area of 'are we testing paginated content or PDF production'. Reftests by definition don't test non-visual differences, so print-reftests should be no different.

I think this is still a valid concern, but it may need slightly reworded in that the risk is that the proposed feature (print-reftests) may not meet the needs of testing authors.